Include/internal/pycore_dict.h

```diff
@@ -246,25 +248,63 @@ _PyDict_NotifyEvent(PyInterpreterState *interp,
     return DICT_NEXT_VERSION(interp) | (mp->ma_version_tag & DICT_WATCHER_AND_MODIFICATION_MASK);
 }
 
-extern PyObject *_PyObject_MakeDictFromInstanceAttributes(PyObject *obj, PyDictValues *values);
-PyAPI_FUNC(bool) _PyObject_MakeInstanceAttributesFromDict(PyObject *obj, PyDictOrValues *dorv);
+extern PyObject *_PyObject_MakeDictFromInstanceAttributes(PyObject *obj);
+
 PyAPI_FUNC(PyObject *)_PyDict_FromItems(
         PyObject *const *keys, Py_ssize_t keys_offset,
         PyObject *const *values, Py_ssize_t values_offset,
         Py_ssize_t length);
 
+static inline uint8_t *
+get_insertion_order_array(PyDictValues *values)
+{
+    return (uint8_t *)&values->values[values->capacity];
+}
+
 static inline void
 _PyDictValues_AddToInsertionOrder(PyDictValues *values, Py_ssize_t ix)
 {
     assert(ix < SHARED_KEYS_MAX_SIZE);
-    uint8_t *size_ptr = ((uint8_t *)values)-2;
-    int size = *size_ptr;
-    assert(size+2 < DICT_VALUES_SIZE(values));
-    size++;
-    size_ptr[-size] = (uint8_t)ix;
-    *size_ptr = size;
+    int size = values->size;
+    uint8_t *array = get_insertion_order_array(values);
+    assert(size < values->capacity);
+    assert(((uint8_t)ix) == ix);
+    array[size] = (uint8_t)ix;
+    values->size = size+1;
+}
+
+static inline size_t
+shared_keys_usable_size(PyDictKeysObject *keys)
+{
+#ifdef Py_GIL_DISABLED
+    // dk_usable will decrease for each instance that is created and each
+    // value that is added.  dk_nentries will increase for each value that
+    // is added.  We want to always return the right value or larger.
+    // We therefore increase dk_nentries first and we decrease dk_usable
+    // second, and conversely here we read dk_usable first and dk_entries
+    // second (to avoid the case where we read entries before the increment
+    // and read usable after the decrement)
+    return (size_t)(_Py_atomic_load_ssize_acquire(&keys->dk_usable) +
+                    _Py_atomic_load_ssize_acquire(&keys->dk_nentries));
+#else
+    return (size_t)keys->dk_nentries + (size_t)keys->dk_usable;
+#endif
```

Comment:
You can add utility macro `FT_ATOMIC_LOAD_SSIZE_ACQUARE` to `pycore_pyatomic_ft_wrappers.h` as same as 
https://github.com/python/cpython/blob/59167c962efcae72e8d88aa4b33062ed3de4f120/Include/internal/pycore_pyatomic_ft_wrappers.h#L22-L38

And remove `ifdef` macro from the logics.

```suggestion
    ssize_t dk_usable = FT_ATOMIC_LOAD_SSIZE_ACQUARE(keys->dk_usable);
    ssize_t dk_nentries = FT_ATOMIC_LOAD_SSIZE_ACQUARE(keys->dk_nentries);
    return dk_nentries + dk_usable;
```

This is what we recommended way while working on PEP 703.

Reply:
That code is moved verbatim from `dictobject.c`, https://github.com/python/cpython/pull/116115/files#diff-b08a47ddc5bc20b2e99ac2e5aa199ca24a56b994e7bc64e918513356088c20aeL892. 
It was introduced by one of the free-threading PRs.

Reply:
Ah okay, so it's not your code.

Reply:
Guessing it was one of @DinoV's recent PRs.

Reply:
I think the `FT_` helpers and the `PyDictKeysObject` thread safety that introduced this were just underdevelopment at the same time, it looks like the `FT_` stuff landed 4 days before this, but it didn't come up in the review so I wasn't aware of it yet.

I'm already building on top of this PR for thread safety of the values and can make this change if @markshannon just wants to leave it as it original was.

Reply:
I'll leave it as it is. That way @DinoV can update the comment appropriately as well.